### PR TITLE
124 add lampe

### PR DESCRIPTION
**Pull request by maho3**
This PR addresses #124 . It consists of the following changes:

- Adds lampe dependencies to setup.cfg and INSTALL.md
- Adds `LampeRunner` in ili/inference/runner_lampe.py, with a configuration similar to the pydelfi and sbi runners. It only does NPE (no SNPE), but can use a wide variety of flows from [zuko](https://zuko.readthedocs.io/en/stable/index.html).
- Adds `LampeNPE` and `LampeEnsemble` to ili/utils/ndes_pt.py, to mirror similar NDE implementations in sbi (with sampling methods and weighting etc.)
- Adds `TorchLoader` to ili/dataloaders. This allows one to simply give LampeRunner custom train/val torch Dataloaders, which is necessary in the GNN example.
- Adds lampe.ipynb in notebooks to demonstrate application. This includes integration with a pytorch_geomentric graph neural network for embeddings.

Still todo:
- [x] Make a configuration-based lampe example
- [x] Cleanup comments
- [x] Add other zuko NDEs 
- [x] Add proposal/prior weighting
- [x] ValidationRunner is hardcoded for old Loaders. Make it work for `TorchLoader`
- [x] Check GPU functionality
- [x] Add tests (Following PR #121)
- [x] Cleanup lampe.ipynb and have it only demonstrate the GNN integration.
- [x] Add documentation to CONFIG.md

### CODE REVIEW

File: CONFIG.md [hunk 1]
```diff
@@ -91,9 +92,9 @@ save_simulated: False  # whether to concatenate the simulated data into x_file a
 You are also welcome to design your own dataloading objects. They will work with NPE/NLE/NRE models so long as they contain the functions: `__len__`, `get_all_data`, and `get_all_parameters`. For SNPE/SNLE/SNRE models, they must also contain the `simulate` and `get_obs_data` functions. See the [_BaseLoader](./ili/dataloaders/loaders.py#L20) template for more details.
 
 ## Training
-There are three available training engines in ltu-ili, SBIRunner and SBIRunnerSequential for `sbi` (PyTorch) models, and PydelfiRunner for `pydelfi` (Tensorflow) models. Each of these engines can be configured from a `yaml`-like configuration file or from iPython initialization as in [tutorial.ipynb](notebooks/tutorial.ipynb).
+There are three available ILI backend for ltu-ili, including `sbi` (PyTorch), `pydelfi` (Tensorflow), and `lampe` (PyTorch). Each of these engines can be configured from a `yaml`-like configuration file or from iPython initialization as in [tutorial.ipynb](notebooks/tutorial.ipynb) or [lampe.ipynb](notebooks/lampe.ipynb).
```
Comment (DeaglanBartlett) [inline CONFIG.md:95]: backend -> backends

File: ili/dataloaders/loaders.py [hunk 2]
```diff
@@ -475,5 +476,71 @@ def load_parameters(
         ).iloc[nodes]
         return theta[param_names].values
 
+
+class TorchLoader(_BaseLoader):
```
Comment (DeaglanBartlett) [inline ili/dataloaders/loaders.py:480]: Please add unit tests to ensure that all functions in this class return what is as expected (as we do for e.g. the numpy loader)

File: ili/inference/runner_lampe.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+        prior = load_from_config(config["prior"])
+
+        # load proposal distributions
+        proposal = None
+        if "proposal" in config:
+            config['proposal']['args']['device'] = config['device']
+            proposal = load_from_config(config["proposal"])
+
+        # load embedding net
+        if "embedding_net" in config:
```
Comment (DeaglanBartlett) [inline ili/inference/runner_lampe.py:109]: Please test this works in the unit tests

**Review by DeaglanBartlett — CHANGES_REQUESTED** [review]
Looks great! Just a couple of minor checks to add to the unit tests and one typo

File: CONFIG.md (hunk 1 above)
Comment (maho3) [reply CONFIG.md:95]: Fixed

File: ili/inference/runner_lampe.py (hunk 3 above)
Comment (maho3) [reply ili/inference/runner_lampe.py:109]: Added.

File: ili/dataloaders/loaders.py (hunk 2 above)
Comment (maho3) [reply ili/dataloaders/loaders.py:480]: Done.

**Pull request merged**